src/lib/snark_worker/prod.ml

```diff
@@ -297,4 +323,105 @@ module Impl = struct
+      ~state:
+        ({ proof_level_snark; cache; proof_cache_db; logger } : Worker_state.t)
+      ~spec:(partitioned_spec : Work.Spec.Partitioned.Stable.Latest.t) :
+      Work.Result.Partitioned.Stable.Latest.t Deferred.Or_error.t =
+    let open Deferred.Or_error.Let_syntax in
+    let sok_digest =
+      Work.Spec.Partitioned.Poly.sok_message partitioned_spec
+      |> Sok_message.digest
+    in
+    let signature_kind = Mina_signature_kind.t_DEPRECATED in
+    match proof_level_snark with
+    | Worker_state.Full ((module M) as m) -> (
+        match partitioned_spec with
+        | Work.Spec.Partitioned.Poly.Single
+            { job = { spec = single_spec; _ } as job; data = () } ->
+            let statement = Work.Spec.Single.Poly.statement single_spec in
+            let%map proof, elapsed =
+              cache_and_time ~logger ~cache ~statement ~single_spec
+                (perform_single_uncached ~m ~logger ~proof_cache_db ~single_spec
+                   ~sok_digest ~signature_kind )
+            in
+            Work.Spec.Partitioned.Poly.Single
+              { job = { job with spec = () }
+              ; data = { Proof_carrying_data.data = elapsed; proof }
+              }
+        | Work.Spec.Partitioned.Poly.Sub_zkapp_command
+            { job =
+                { spec =
+                    Work.Spec.Sub_zkapp.Stable.Latest.Segment
+                      { statement; witness; spec = segment_spec; _ } as
+                    sub_zkapp_spec
+                ; _
+                } as job
+            ; data = ()
+            } ->
+            let witness =
+              Transaction_witness.Zkapp_command_segment_witness
+              .write_all_proofs_to_disk ~proof_cache_db witness
+            in
+
+            let statement_without_sok =
+              Work.Spec.Sub_zkapp.Stable.Latest.statement sub_zkapp_spec
+            in
+
+            let%map proof, elapsed =
+              log_subzkapp_base_snark ~logger ~statement ~spec:segment_spec
+                (M.of_zkapp_command_segment_exn ~witness)
+              |> cache_and_time_partitioned ~logger ~cache
+                   ~statement:statement_without_sok ~partitioned_spec
+            in
+
+            Work.Spec.Partitioned.Poly.Sub_zkapp_command
+              { job = { job with spec = () }
+              ; data = { Proof_carrying_data.data = elapsed; proof }
+              }
+        | Work.Spec.Partitioned.Poly.Sub_zkapp_command
+            { job =
+                { spec =
+                    Work.Spec.Sub_zkapp.Stable.Latest.Merge { proof1; proof2 }
+                    as spec
+                ; _
+                } as job
+            ; data = ()
+            } ->
+            let statement_without_sok =
+              Work.Spec.Sub_zkapp.Stable.Latest.statement spec
+            in
+
+            let%map proof, elapsed =
+              log_subzkapp_merge_snark ~m ~logger ~sok_digest proof1 proof2
+              |> cache_and_time_partitioned ~logger ~cache
```

Comment:
I'd suggest not using cache for the new code, I see no value i having it in the first place.